projection/validator_delegation/delegation.go

```diff
@@ -0,0 +1,269 @@
+	}
+	if !found {
+		return coin.ZeroDec(), fmt.Errorf("error validator not found: %v, %v", validatorAddress, height)
+	}
+
+	// Get the delegation
+	delegation, found, err := delegationsView.FindBy(
+		validatorAddress,
+		delegatorAddress,
+		height,
+	)
+	if err != nil {
+		return coin.ZeroDec(), fmt.Errorf("error delegationsView.FindBy(): %v", err)
+	}
+	if !found {
+		return coin.ZeroDec(), fmt.Errorf("error delegation not found: %v, %v, %v", validatorAddress, delegatorAddress, height)
+	}
+
+	var shares coin.Dec
+	shares, err = utils.ValidatorSharesFromTokens(validator, amount)
+	if err != nil {
+		return coin.ZeroDec(), fmt.Errorf("error utils.ValidatorSharesFromTokens(): %v", err)
+	}
+	// Cap the shares at the delegation's shares.
+	if shares.GT(delegation.Shares) {
+		shares = delegation.Shares
+	}
+
+	return shares, nil
+}
+
+func (projection *ValidatorDelegation) unbond(
+	rdbTxHandle *rdb.Handle,
+	height int64,
+	validatorAddress string,
+	delegatorAddress string,
+	shares coin.Dec,
+) (coin.Int, error) {
+
+	validatorsView := NewValidators(rdbTxHandle)
+	delegationsView := NewDelegations(rdbTxHandle)
+
+	// Get the validator
+	validator, found, err := validatorsView.FindByOperatorAddr(validatorAddress, height)
+	if err != nil {
+		return coin.ZeroInt(), fmt.Errorf("error validatorsView.FindByOperatorAddr(): %v", err)
+	}
+	if !found {
+		return coin.ZeroInt(), fmt.Errorf("error validator not found: %v, %v", validatorAddress, height)
+	}
+
+	// Get the delegation
+	delegation, found, err := delegationsView.FindBy(
+		validatorAddress,
+		delegatorAddress,
+		height,
+	)
+	if err != nil {
+		return coin.ZeroInt(), fmt.Errorf("error delegationsView.FindBy(): %v", err)
+	}
+	if !found {
+		return coin.ZeroInt(), fmt.Errorf("error delegation not found: %v, %v, %v", validatorAddress, delegatorAddress, height)
+	}
+
+	// Subtract shares from delegation
+	delegation.Shares = delegation.Shares.Sub(shares)
+
+	// Check if this delegation is the validator self delegation
+	isOperatorDelegation, err := utils.IsValAddrEqualsDelAddr(
+		validatorAddress,
+		delegatorAddress,
+		projection.config.validatorAddressPrefix,
+		projection.config.accountAddressPrefix,
+	)
+	if err != nil {
+		return coin.ZeroInt(), fmt.Errorf("error in IsValAddrEqualsDelAddr: %v", err)
+	}
+
+	// Remaining tokens in the delegation
+	delegationRemainingTokens := utils.ValidatorTokensFromShares(validator, delegation.Shares)
+
+	// If the delegation belongs to operator of the validator and undelegating will decrease the validator's
+	// self-delegation below their minimum, we jail the validator.
+	if isOperatorDelegation && !validator.Jailed &&
+		delegationRemainingTokens.TruncateInt().LT(validator.MinSelfDelegation) {
+
+		if err := projection.jailValidator(rdbTxHandle, validator); err != nil {
+			return coin.ZeroInt(), fmt.Errorf("error in projection.jailValidator(): %v", err)
+		}
+		validator.Jailed = true
```

Comment:
So there are two separate fields store the `Status` (bonded/unbonding/unbonded) and `Jailed` (bool).

In this case, the `Validator.Jailed` will immediately be set to true. And there is a `PowerIndex` (in Cosmos), validator will be removed from the store immediately.
https://github.com/cosmos/cosmos-sdk/blob/9aef070625e9676d7c0acb212c17ae9dba3c32dc/x/staking/keeper/delegation.go#L675
https://github.com/cosmos/cosmos-sdk/blob/9aef070625e9676d7c0acb212c17ae9dba3c32dc/x/staking/keeper/val_state_change.go#L264

Then later, in the EndBlock section, the `Validator.Status` will be updated to `unbonding`.
 https://github.com/cosmos/cosmos-sdk/blob/9aef070625e9676d7c0acb212c17ae9dba3c32dc/x/staking/keeper/val_state_change.go#L112
 (check line 122, 128, 131, 183, 189, 194)
 
 To answer the question, after the block finishes everything, the `Validator.Status = unbonding` and `Validator.Jailed = true`.

Reply:
Does this jail involve a jailing period? Or it's just a status update?

Because usually when we say a validator jailed it means the validator cannot perform any action during jailing period.

Reply:
No, in this case, there is no jailing period. If the validator delegate enough tokens to itself, it can be immediately unjailed through a MsgUnjail.

Here is a good reference to the question: https://github.com/cosmos/cosmos-sdk/blob/2f335a2b8cc4fcfa337c104823fab9f6ecb387da/x/slashing/keeper/unjail.go#L40

I think the jailing period you mentioned relates to `ValidatorSigningInfo`, there is a `JailedUntil` field under it. But `JailedUntil`seems only in use when the jail is due to `missing_signature`.

`ValidatorSigningInfo` link: https://github.com/cosmos/cosmos-sdk/blob/2f335a2b8cc4fcfa337c104823fab9f6ecb387da/x/slashing/types/slashing.pb.go#L33